XiEditor/SidebarViewController.swift

```diff
@@ -0,0 +1,165 @@
+            } else {
+                sidebar.appearance = NSAppearance(named: .vibrantDark)
+            }
+        } else {
+            sidebar.appearance = NSAppearance(named: .aqua)
+        }
+        sidebar.backgroundColor = self.theme.background
+    }
+
+    /// TODO: Sync the structure (expanded/collapsed/selected) of each windows sidebar
+    public func syncStructure() {
+        
+    }
+
+    /// Called when the a row in the sidebar is double clicked
+    @objc private func doubleClicked(_ sender: Any?) {
+        let clickedRow = sidebar.item(atRow: sidebar.clickedRow)
+
+        if sidebar.isItemExpanded(clickedRow) {
+            sidebar.collapseItem(clickedRow)
+        } else {
+            sidebar.expandItem(clickedRow)
+        }
+    }
+
+    @objc private func updateItems() {
+        self.sidebar.reloadData()
+    }
+
+}
+
+// MARK: - Sidebar NSOutlineViewDataSource
+
+extension SidebarViewController: NSOutlineViewDataSource {
+
+    // Number of items in the sidebar
+    func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
+        guard let item = item as? FileSystemItem else { return sidebarItems.items.count }
+        return item.getNumberOfChildren()
+    }
+
+    // Items to be added to sidebar
+    func outlineView(_ outlineView: NSOutlineView, child index: Int, ofItem item: Any?) -> Any {
+        guard let item = item as? FileSystemItem else { return sidebarItems.items[index] }
+        return item.getChild(at: index)
+    }
+
+    // Whether rows are expandable by an arrow
+    func outlineView(_ outlineView: NSOutlineView, isItemExpandable item: Any) -> Bool {
+        guard let item = item as? FileSystemItem else { return false }
+        return item.getNumberOfChildren() != 0
+    }
+
+    // Height of each row
+    func outlineView(_ outlineView: NSOutlineView, heightOfRowByItem item: Any) -> CGFloat {
+        return 20.0
+    }
+
+    // When a row is clicked on should it be selected
+    func outlineView(_ outlineView: NSOutlineView, shouldSelectItem item: Any) -> Bool {
+        return true
+    }
+
+    // When a row is selected
+    func outlineViewSelectionDidChange(_ notification: Notification) {
+        guard
+            let outlineView = notification.object as? NSOutlineView,
+            let doc = outlineView.item(atRow: outlineView.selectedRow) as? FileSystemItem
+        else { return }
+
+        let rows = IndexSet(integersIn: 0..<outlineView.numberOfRows)
+
+        rows.compactMap { outlineView.rowView(atRow: $0, makeIfNecessary: false) }
+            .forEach { $0.backgroundColor = $0.isSelected ? .selectedControlColor : .clear }
+
+        if doc.isDirectory { return }
+
+        // Get the document controller and open the selected document into a new tab
+        if let appDelegate = NSApplication.shared.delegate as? AppDelegate {
+            appDelegate.documentController.openDocumentIntoNewTab(
```

Comment:
Just UX stuff, but I think the behavior should be if it's a single click, open in current window and if it's a double click, open in new tab. `doubleAction` should work here.

Reply:
@nangtrongvuon I thought this would be a good idea too, similar to how Xcode handles it. Although, how would we account for unsaved file changes to a document? The way Xcode does it is that it doesn't save the file and stores the changes in the background, so when the file is reopened, the changes are still there. And if a file has unsaved changes when Xcode quits, the changes get saved automatically. Does xi-editor support this kind of behavior?